doc/man/gnunet-vpn.1
.TH GNUNET-VPN "1" "10 Dec 2007" "GNUnet"
.SH NAME
gnunet\-vpn \- maintain your GNUbone connection
.SH SYNOPSIS
.B gnunet\-vpn
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
gnunet\-vpn is a tool to be used together with the GNUnet VPN module.
See http://uk.gnunet.org/vpn/ for more information about the VPN module.
.SH USAGE

These are the more interesting commands available in gnunet-vpn. For normal operation you would invoke "reset", and "realise", alternately every hour or so. To give you something to access, you may do "add " so long as you have this node?s hello files in your cache, that can be found in my hosts area. Then try ping6 vpn1.wensley.org.uk, if address unreachable comes back, ping6 that instead. I do have an ftp and gopher server accessible here. 

.TP
\fBreset\fR
Clears prototype routing table and start requesting routes from peers to rebuild it
.TP
\fBrealise\fR
Realise prototype table into the real routing table, and adds and pulls IPv6 routes to match it. You use this when you think your prototype table is complete. I.e. peers have stopped giving you more routes.
.TP
\fBtunnel\fR
Show tunnels to peers
\fBroutes\fR
Show prototype routing table contents
.TP
\fBrealised\fR
Show real routing table contents
.TP
\fbadd \fIHASH\fR
Persuade GNUnet to connect to the peer with the given hash.
.TP
\fBdebug0\fR
Hide activity output.
\fBdebug1\fR
Show activity output.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.TP
\fB\-s\fR, \fB\-\-silent\fR
run silently
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number

.SH AUTHOR
Michael John Wensley
.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/vpn/gnunet-vpn.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2004, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/vpn/gnunet-vpn.c
 * @brief Utility to admin VPN
 * @author Michael John Wensley
 *
 * TODO:
 * - clean up use of semaphores / signaling
 * - make proper use of structs
 */

#include "gnunet_util.h"
#include "gnunet_protocols.h"
#include "gnunet_directories.h"
#include "platform.h"

#define TEMPLATE_VERSION "2006072900"

#define buf ((GNUNET_MessageHeader*)&buffer)

static struct GNUNET_Semaphore *doneSem;

static struct GNUNET_Semaphore *cmdAck;

static struct GNUNET_Semaphore *exitCheck;

static struct GNUNET_Mutex *lock;

static int wantExit;

static int silent;

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

/**
 * All gnunet-transport-check command line options
 */
static struct GNUNET_CommandLineOption gnunetvpnOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Print statistics about GNUnet operations.")), /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'s', "silent", NULL,
   gettext_noop ("Suppress display of asynchronous log messages"),
   0, &GNUNET_getopt_configure_set_one, &silent},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_END,
};

static void *
receiveThread (void *arg)
{
  struct GNUNET_ClientServerConnection *sock = arg;
  char buffer[GNUNET_MAX_BUFFER_SIZE];
  GNUNET_MessageHeader *bufp = buf;

  while (GNUNET_OK == GNUNET_client_connection_read (sock, &bufp))
    {
      switch (ntohs (buf->type))
        {
        case GNUNET_CS_PROTO_VPN_TUNNELS:
        case GNUNET_CS_PROTO_VPN_ROUTES:
        case GNUNET_CS_PROTO_VPN_REALISED:
        case GNUNET_CS_PROTO_VPN_RESET:
        case GNUNET_CS_PROTO_VPN_ADD:
        case GNUNET_CS_PROTO_VPN_TRUST:
          if (ntohs (buf->size) > sizeof (GNUNET_MessageHeader))
            {
              fwrite (buffer + sizeof (GNUNET_MessageHeader),
                      sizeof (char),
                      ntohs (buf->size) - sizeof (GNUNET_MessageHeader),
                      stdout);
            }

          GNUNET_semaphore_up (cmdAck);
          GNUNET_semaphore_down (exitCheck, GNUNET_YES);
          GNUNET_mutex_lock (lock);
          if (wantExit == GNUNET_YES)
            {
              GNUNET_mutex_unlock (lock);
              GNUNET_semaphore_up (doneSem);
              return NULL;
            }
          GNUNET_mutex_unlock (lock);
          break;
        case GNUNET_CS_PROTO_VPN_MSG:
          if (silent == GNUNET_YES)
            break;
        case GNUNET_CS_PROTO_VPN_REPLY:

          if (ntohs (buf->size) > sizeof (GNUNET_MessageHeader))
            {
              fwrite (buffer + sizeof (GNUNET_MessageHeader),
                      sizeof (char),
                      ntohs (buf->size) - sizeof (GNUNET_MessageHeader),
                      stdout);
            }
          break;;
        }
    }
  GNUNET_semaphore_up (doneSem);
  return NULL;
}

#define COMMAND_LINE_SIZE 1024

/**
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunet-template: 0: ok, -1: error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_ClientServerConnection *sock;
  struct GNUNET_ThreadHandle *messageReceiveThread;
  void *unused;
  char buffer[sizeof (GNUNET_MessageHeader) + COMMAND_LINE_SIZE];
  int rancommand = 0;
  struct GNUNET_GC_Configuration *cfg;
  struct GNUNET_GE_Context *ectx;
  int i;

  i = GNUNET_init (argc,
                   argv,
                   "gnunet-vpn", &cfgFilename, gnunetvpnOptions, &ectx, &cfg);
  if (i == -1)
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  sock = GNUNET_client_connection_create (ectx, cfg);
  if (sock == NULL)
    {
      fprintf (stderr, _("Error establishing connection with gnunetd.\n"));
      GNUNET_fini (ectx, cfg);
      return 1;
    }

  doneSem = GNUNET_semaphore_create (0);
  cmdAck = GNUNET_semaphore_create (0);
  exitCheck = GNUNET_semaphore_create (0);
  lock = GNUNET_mutex_create (GNUNET_NO);
  wantExit = GNUNET_NO;

  messageReceiveThread =
    GNUNET_thread_create (&receiveThread, sock, 128 * 1024);
  if (messageReceiveThread == NULL)
    GNUNET_GE_DIE_STRERROR (ectx,
                            GNUNET_GE_FATAL | GNUNET_GE_ADMIN | GNUNET_GE_USER
                            | GNUNET_GE_IMMEDIATE, "pthread_create");


  /* accept keystrokes from user and send to gnunetd */
  while (NULL != fgets (buffer, COMMAND_LINE_SIZE, stdin))
    {
      if (rancommand)
        {
          rancommand = 0;
          GNUNET_semaphore_up (exitCheck);
        }
      else if (strncmp (buffer, "tunnels", 7) == 0)
        {
          ((GNUNET_MessageHeader *) & buffer)->type =
            htons (GNUNET_CS_PROTO_VPN_TUNNELS);
          ((GNUNET_MessageHeader *) & buffer)->size =
            htons (sizeof (GNUNET_MessageHeader));
          if (GNUNET_SYSERR ==
              GNUNET_client_connection_write (sock,
                                              (GNUNET_MessageHeader *) &
                                              buffer))
            return -1;
          rancommand = 1;
          GNUNET_semaphore_down (cmdAck, GNUNET_YES);
        }
      else if (strncmp (buffer, "route", 5) == 0)
        {
          ((GNUNET_MessageHeader *) & buffer)->type =
            htons (GNUNET_CS_PROTO_VPN_ROUTES);
          ((GNUNET_MessageHeader *) & buffer)->size =
            htons (sizeof (GNUNET_MessageHeader));
          if (GNUNET_SYSERR ==
              GNUNET_client_connection_write (sock,
                                              (GNUNET_MessageHeader *) &
                                              buffer))
            return -1;
          rancommand = 1;
          GNUNET_semaphore_down (cmdAck, GNUNET_YES);
        }
      else if (strncmp (buffer, "realised", 8) == 0)
        {
          ((GNUNET_MessageHeader *) & buffer)->type =
            htons (GNUNET_CS_PROTO_VPN_REALISED);
          ((GNUNET_MessageHeader *) & buffer)->size =
            htons (sizeof (GNUNET_MessageHeader));
          if (GNUNET_SYSERR ==
              GNUNET_client_connection_write (sock,
                                              (GNUNET_MessageHeader *) &
                                              buffer))
            return -1;
          rancommand = 1;
          GNUNET_semaphore_down (cmdAck, GNUNET_YES);
        }
      else if (strncmp (buffer, "reset", 5) == 0)
        {
          ((GNUNET_MessageHeader *) & buffer)->type =
            htons (GNUNET_CS_PROTO_VPN_RESET);
          ((GNUNET_MessageHeader *) & buffer)->size =
            htons (sizeof (GNUNET_MessageHeader));
          if (GNUNET_SYSERR ==
              GNUNET_client_connection_write (sock,
                                              (GNUNET_MessageHeader *) &
                                              buffer))
            return -1;
          rancommand = 1;
          GNUNET_semaphore_down (cmdAck, GNUNET_YES);
        }
      else if (strncmp (buffer, "trust", 5) == 0)
        {
          ((GNUNET_MessageHeader *) & buffer)->type =
            htons (GNUNET_CS_PROTO_VPN_TRUST);
          ((GNUNET_MessageHeader *) & buffer)->size =
            htons (sizeof (GNUNET_MessageHeader));
          if (GNUNET_SYSERR ==
              GNUNET_client_connection_write (sock,
                                              (GNUNET_MessageHeader *) &
                                              buffer))
            return -1;
          rancommand = 1;
          GNUNET_semaphore_down (cmdAck, GNUNET_YES);
        }
      else if (strncmp (buffer, "add ", 4) == 0)
        {
          /* message header is 4 bytes long, we overwrite "add " with it
           * also don't include \r or \n in the message
           */
          if (strlen (&buffer[4]) > 1)
            {
              ((GNUNET_MessageHeader *) & buffer)->type =
                htons (GNUNET_CS_PROTO_VPN_ADD);
              ((GNUNET_MessageHeader *) & buffer)->size =
                htons (sizeof (GNUNET_MessageHeader) + strlen (&buffer[5]));
              if (GNUNET_SYSERR ==
                  GNUNET_client_connection_write (sock,
                                                  (GNUNET_MessageHeader *) &
                                                  buffer))
                return -1;
              rancommand = 1;
              GNUNET_semaphore_down (cmdAck, GNUNET_YES);
            }
          else
            {
              printf ("add requires hash as a parameter!\n");
            }
        }
      else
        {
          printf ("tunnels, route, realised, reset, trust, add <hash>\n");
        }
    }
  /* wait for shutdown... */
  if (rancommand)
    {
      GNUNET_mutex_lock (lock);
      wantExit = GNUNET_YES;
      GNUNET_mutex_unlock (lock);
      GNUNET_semaphore_up (exitCheck);
    }

  /* we can't guarantee that this can be called while the other thread is waiting for read */
  GNUNET_client_connection_close_forever (sock);
  GNUNET_semaphore_down (doneSem, GNUNET_YES);
  GNUNET_semaphore_destroy (doneSem);
  GNUNET_semaphore_destroy (cmdAck);
  GNUNET_semaphore_destroy (exitCheck);
  GNUNET_mutex_destroy (lock);
  GNUNET_thread_join (messageReceiveThread, &unused);
  GNUNET_client_connection_destroy (sock);
  GNUNET_fini (ectx, cfg);

  return 0;
}

/* end of gnunet-vpn.c */
